### PR TITLE
[DML] QAttention

### PR DESCRIPTION
**Pull request by raoanag**
### Description
DML Implementation for [com.microsoft.QAttention](https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.QAttention)

### CODE REVIEW

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorQAttention.cpp [hunk 1]
```diff
@@ -0,0 +1,703 @@
...
+// Licensed under the MIT License.
+
+#include "precomp.h"
+
+/*
+Abbreviations: B is batch_size, S is sequence_length, W is hidden_size
+               N is number of attention heads, H is head size, and W=N*H
+               M is mask_index tensor
+
+M, A, B, C and P are Inputs
```
Comment (sumitsays) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorQAttention.cpp:11]: >M [](http://example.com/codeflow?start=0&length=1)

Now that we are handling few extra cases, where M can be generated on the fly using Diagonal_Matrix1, it is not always true that `M` will be an input provided by the model.
Can we please remove `M` from this comment line and add a separate comment line explaining that `M` can be a input or can be generated during runtime?

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp [hunk 2]
```diff
@@ -613,15 +615,23 @@ constexpr static std::array<SupportedTensorDataTypes, 2> supportedTypeListLayerN
 constexpr static std::array<SupportedTensorDataTypes, 2> supportedTypeListShape = {SupportedTensorDataTypes::All, SupportedTensorDataTypes::Int64};
 constexpr static std::array<SupportedTensorDataTypes, 2> supportedTypeListSize = {SupportedTensorDataTypes::All, SupportedTensorDataTypes::Int64};
 constexpr static std::array<SupportedTensorDataTypes, 1> supportedTypeListQLinearSigmoid = {SupportedTensorDataTypes::UInt8 | SupportedTensorDataTypes::Int8};
+
+constexpr static std::array<SupportedTensorDataTypes, 4> supportedTypeListQAttention = {
+    SupportedTensorDataTypes::Ints8Bit,
+    SupportedTensorDataTypes::Ints8Bit,
+    SupportedTensorDataTypes::Float16to32,
+    SupportedTensorDataTypes::Int32
+};
```
Comment (sumitsays) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp:624]: [nit] we should keep the same style of having this array in single line to make it consistent.

Comment (raoanag) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/OperatorRegistration.cpp:624]: I feel this is more easy to understand individual input data type, there seems to be a mix of single and multiple line. Multiple lines used for input count >2

**Review by sumitsays — APPROVED** [review]
:shipit:

**Review by sumitsays — APPROVED** [review]

**Pull request merged**